src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Instance.hs

```diff
@@ -163,19 +167,31 @@ instance {-# OVERLAPPING #-} (SymbolicData a x, n ~ TypeSize a x, KnownNat n) =>
                 bs <- V.item <$> runCircuit b
                 V.zipWithM (\x y -> newAssigned $ \p -> p bs * (p x - p y) + p y) ts fs
 
--- TODO: make a proper implementation of Arbitrary
-instance (Arithmetic a, KnownNat n) => Arbitrary (ArithmeticCircuit n a) where
-    arbitrary = return $ c1 * c2
-        where
-            c1, c2 :: ArithmeticCircuit n a
-            c1 = ArithmeticCircuit { acCircuit = mempty, acOutput = pure 1}
-            c2 = ArithmeticCircuit { acCircuit = mempty, acOutput = pure 2}
-
+instance (Arithmetic a, Arbitrary a) => Arbitrary (ArithmeticCircuit 1 a) where
+    arbitrary = do
+        k <- integerToNatural <$> chooseInteger (2, 10)
+        let ac = ArithmeticCircuit { acCircuit = mempty {acInput = [1..k]}, acOutput = pure k }
+        arbitrary' ac 10
+
+arbitrary' :: forall a . (Arithmetic a, Arbitrary a, FromConstant a a) => ArithmeticCircuit 1 a -> Natural -> Gen (ArithmeticCircuit 1 a)
+arbitrary' ac 0 = return ac
+arbitrary' ac iter = do
+    let vars = getAllVars . acCircuit $ ac
+    li <- oneof $ fmap return vars
```

Comment:
[`elements`](https://hackage.haskell.org/package/QuickCheck-2.15.0.1/docs/Test-QuickCheck.html#v:elements) can be used instead of `oneof` in situations like this